### DISCUSSION TITLE
Negative pickup and demand capacity issue

### DISCUSSION THREAD

**Discussion by SergiBago** [google/or-tools#3074] [Routing (and legacy CP) questions]
Good morning,

I have to solve a problem, in which a company has different jobs in diferent places of a city. The idea is that instead of each worker having his own car, the company puts up a van that picks up and drops off workers. All workers have the same functions, so anyone can go to any job. In addition, the jobs have specific time windows.

So, I have done an example of time window with capacity constrains, where the pickups are positive values and the dropoffs are negative values, the only problem is that I have negative charges. That is, the vehicle tries to leave more workers than it carries. With the upper limit of the loads I have no problem.

I would like to know how I can do to restrict that the minimum of the load in the vehicles has to be 0.

Code:
```py
!pip install ortools

"""Vehicles Routing Problem (VRP) with Time Windows."""

from ortools.constraint_solver import routing_enums_pb2
from ortools.constraint_solver import pywrapcp


def create_data_model():
    """Stores the data for the problem."""
    #TIME PART
    data = {}
    data['time_matrix'] = [
        [0, 6, 9, 8, 7, 3, 6, 2, 3, 2, 6, 6, 4, 4, 5, 9, 7],
        [6, 0, 8, 3, 2, 6, 8, 4, 8, 8, 13, 7, 5, 8, 12, 10, 14],
        [9, 8, 0, 11, 10, 6, 3, 9, 5, 8, 4, 15, 14, 13, 9, 18, 9],
        [8, 3, 11, 0, 1, 7, 10, 6, 10, 10, 14, 6, 7, 9, 14, 6, 16],
        [7, 2, 10, 1, 0, 6, 9, 4, 8, 9, 13, 4, 6, 8, 12, 8, 14],
        [3, 6, 6, 7, 6, 0, 2, 3, 2, 2, 7, 9, 7, 7, 6, 12, 8],
        [6, 8, 3, 10, 9, 2, 0, 6, 2, 5, 4, 12, 10, 10, 6, 15, 5],
        [2, 4, 9, 6, 4, 3, 6, 0, 4, 4, 8, 5, 4, 3, 7, 8, 10],
        [3, 8, 5, 10, 8, 2, 2, 4, 0, 3, 4, 9, 8, 7, 3, 13, 6],
        [2, 8, 8, 10, 9, 2, 5, 4, 3, 0, 4, 6, 5, 4, 3, 9, 5],
        [6, 13, 4, 14, 13, 7, 4, 8, 4, 4, 0, 10, 9, 8, 4, 13, 4],
        [6, 7, 15, 6, 4, 9, 12, 5, 9, 6, 10, 0, 1, 3, 7, 3, 10],
        [4, 5, 14, 7, 6, 7, 10, 4, 8, 5, 9, 1, 0, 2, 6, 4, 8],
        [4, 8, 13, 9, 8, 7, 10, 3, 7, 4, 8, 3, 2, 0, 4, 5, 6],
        [5, 12, 9, 14, 12, 6, 6, 7, 3, 3, 4, 7, 6, 4, 0, 9, 2],
        [9, 10, 18, 6, 8, 12, 15, 8, 13, 9, 13, 3, 4, 5, 9, 0, 9],
        [7, 14, 9, 16, 14, 8, 5, 10, 6, 5, 4, 10, 8, 6, 2, 9, 0],
    ]
    data['time_windows'] = [
        (0, 5),  # depot
        (7, 12),  # 1
        (10, 15),  # 2
        (16, 18),  # 3
        (10, 13),  # 4
        (0, 5),  # 5
        (5, 10),  # 6
        (0, 4),  # 7
        (5, 10),  # 8
        (0, 3),  # 9
        (10, 16),  # 10
        (10, 15),  # 11
        (0, 5),  # 12
        (5, 10),  # 13
        (7, 8),  # 14
        (10, 15),  # 15
        (11, 15),  # 16
    ]
    data['num_vehicles'] = 6   
    data['depot'] = 0

    # PICKUP AND DELIVERIES

    #PICKUPS
    data['demand'] = [0, 1, 1, -2, 4, 2, -4, 8, 8, 1, -2, 1, -2, 4, 4, -8, 8]

    data['vehicle_capacities'] = [15, 15, 15, 15]


    return data


def print_solution(data, manager, routing, solution):
    """Prints solution on console."""
    print(f'Objective: {solution.ObjectiveValue()}')
    time_dimension = routing.GetDimensionOrDie('Time')
    total_time = 0
    total_load = 0
    for vehicle_id in range(data['num_vehicles']):
        index = routing.Start(vehicle_id)
        plan_output = 'Route for vehicle {}:\n'.format(vehicle_id)
        route_load = 0
        while not routing.IsEnd(index):
            time_var = time_dimension.CumulVar(index)
            node_index = manager.IndexToNode(index)
            route_load += data['demand'][node_index]

            plan_output += '{0} Time({1},{2}), Load {3} -> '.format( manager.IndexToNode(index), solution.Min(time_var), solution.Max(time_var), route_load)
            index = solution.Value(routing.NextVar(index))
        time_var = time_dimension.CumulVar(index)
        plan_output += '{0} Time({1},{2}), Load {3}\n'.format(manager.IndexToNode(index),
                                                    solution.Min(time_var),
                                                    solution.Max(time_var), route_load)
        plan_output += 'Time of the route: {}min\n'.format(
            solution.Min(time_var))
        print(plan_output)
        total_time += solution.Min(time_var)
    print('Total time of all routes: {}min'.format(total_time))


def main():
    """Solve the VRP with time windows."""
    # Instantiate the data problem.
    data = create_data_model()

    # Create the routing index manager.
    manager = pywrapcp.RoutingIndexManager(len(data['time_matrix']),
                                           data['num_vehicles'], data['depot'])

    # Create Routing Model.
    routing = pywrapcp.RoutingModel(manager)

    ##### ---- TIME ---- 
    # Create and register a transit callback.
    def time_callback(from_index, to_index):
        """Returns the travel time between the two nodes."""
        # Convert from routing variable Index to time matrix NodeIndex.
        from_node = manager.IndexToNode(from_index)
        to_node = manager.IndexToNode(to_index)
        return data['time_matrix'][from_node][to_node]

    transit_callback_index = routing.RegisterTransitCallback(time_callback)

    # Define cost of each arc.
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)





    # Add Time Windows constraint.
    time = 'Time'
    routing.AddDimension(
        transit_callback_index,
        30,  # allow waiting time
        30,  # maximum time per vehicle
        False,  # Don't force start cumul to zero.
        time)
    

    time_dimension = routing.GetDimensionOrDie(time)
    # Add time window constraints for each location except depot.
    for location_idx, time_window in enumerate(data['time_windows']):
        if location_idx == data['depot']:
            continue
        index = manager.NodeToIndex(location_idx)
        time_dimension.CumulVar(index).SetRange(time_window[0], time_window[1])
  
    # Add time window constraints for each vehicle start node.
    depot_idx = data['depot']
    for vehicle_id in range(data['num_vehicles']):
        index = routing.Start(vehicle_id)
        time_dimension.CumulVar(index).SetRange(
            data['time_windows'][depot_idx][0],
            data['time_windows'][depot_idx][1])

    # Instantiate route start and end times to produce feasible times.
    for i in range(data['num_vehicles']):
        routing.AddVariableMinimizedByFinalizer(
            time_dimension.CumulVar(routing.Start(i)))
        routing.AddVariableMinimizedByFinalizer(
            time_dimension.CumulVar(routing.End(i)))


    ### PICKUPS AND
    def demand_callback(from_index):
        """Returns the demand of the node."""
        # Convert from routing variable Index to demands NodeIndex.
        from_node = manager.IndexToNode(from_index)
        return data['demand'][from_node]


    demand_callback_index = routing.RegisterUnaryTransitCallback(
        demand_callback)


    capacity= 'Capacity'    
    routing.AddDimension(
        demand_callback_index,
        0,  # null capacity slack
        15,  # vehicle maximum capacities
        False,  # start cumul to zero
        capacity)


    # Setting first solution heuristic.
    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
    search_parameters.first_solution_strategy = (
        routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)

    # Solve the problem.
    solution = routing.SolveWithParameters(search_parameters)

    # Print solution on console.
    if solution:
        print_solution(data, manager, routing, solution)


if __name__ == '__main__':
    main()
```

Solution:

Route for vehicle 0:
0 Time(0,0), Load 0 -> 0 Time(0,0), Load 0
Time of the route: 0min

Route for vehicle 1:
0 Time(0,0), Load 0 -> 0 Time(0,0), Load 0
Time of the route: 0min

Route for vehicle 2:
0 Time(0,0), Load 0 -> 9 Time(2,3), Load 1 -> 14 Time(7,8), Load 5 -> 16 Time(11,11), Load 13 -> 0 Time(18,18), Load 13
Time of the route: 18min

Route for vehicle 3:
0 Time(0,0), Load 0 -> 7 Time(2,4), Load 8 -> 1 Time(7,11), Load 9 -> 4 Time(10,13), Load 13 -> 3 Time(16,16), Load 11 -> 0 Time(24,24), Load 11
Time of the route: 24min

Route for vehicle 4:
0 Time(0,0), Load 0 -> 12 Time(4,4), Load -2 -> 13 Time(6,6), Load 2 -> 15 Time(11,11), Load -6 -> 11 Time(14,14), Load -5 -> 0 Time(20,20), Load -5
Time of the route: 20min

Route for vehicle 5:
0 Time(0,0), Load 0 -> 5 Time(3,3), Load 2 -> 8 Time(5,5), Load 10 -> 6 Time(7,7), Load 6 -> 2 Time(10,10), Load 7 -> 10 Time(14,14), Load 5 -> 0 Time(20,20), Load 5
Time of the route: 20min

Total time of all routes: 82min

**Comment by jmarca** [comment]
Maybe this part is the issue:

```
   capacity= 'Capacity'
   routing.AddDimension(
       demand_callback_index,
       0,  # null capacity slack
       15,  # vehicle maximum capacities
       False,  # start cumul to zero
       capacity)
```

Try setting the initial cumul to zero.

I learned the hard way that dimensions have an absolute floor at zero,
the only way you can drop off more than you pick up is if you let the
solver find its own floor with that `False` in the dimension
initialization command.

Or perhaps it is something else...

James

On Thu, Jan 13, 2022 at 04:48:41AM -0800, Sergi Bagó Vila wrote:
> Good morning,
>
> I have to solve a problem, in which a company has different jobs in diferent places of a city. The idea is that instead of each worker having his own car, the company puts up a van that picks up and drops off workers. All workers have the same functions, so anyone can go to any job. In addition, the jobs have specific time windows.
>
> So, I have done an example of time window with capacity constrains, where the pickups are positive values and the dropoffs are negative values, the only problem is that I have negative charges. That is, the vehicle tries to leave more workers than it carries. With the upper limit of the loads I have no problem.
>
> I would like to know how I can do to restrict that the minimum of the load in the vehicles has to be 0.
>
> Code:
>
> !pip install ortools
>
> """Vehicles Routing Problem (VRP) with Time Windows."""
>
> from ortools.constraint_solver import routing_enums_pb2
> from ortools.constraint_solver import pywrapcp
>
>
> def create_data_model():
>     """Stores the data for the problem."""
>     #TIME PART
>     data = {}
>     data['time_matrix'] = [
>         [0, 6, 9, 8, 7, 3, 6, 2, 3, 2, 6, 6, 4, 4, 5, 9, 7],
>         [6, 0, 8, 3, 2, 6, 8, 4, 8, 8, 13, 7, 5, 8, 12, 10, 14],
>         [9, 8, 0, 11, 10, 6, 3, 9, 5, 8, 4, 15, 14, 13, 9, 18, 9],
>         [8, 3, 11, 0, 1, 7, 10, 6, 10, 10, 14, 6, 7, 9, 14, 6, 16],
>         [7, 2, 10, 1, 0, 6, 9, 4, 8, 9, 13, 4, 6, 8, 12, 8, 14],
>         [3, 6, 6, 7, 6, 0, 2, 3, 2, 2, 7, 9, 7, 7, 6, 12, 8],
>         [6, 8, 3, 10, 9, 2, 0, 6, 2, 5, 4, 12, 10, 10, 6, 15, 5],
>         [2, 4, 9, 6, 4, 3, 6, 0, 4, 4, 8, 5, 4, 3, 7, 8, 10],
>         [3, 8, 5, 10, 8, 2, 2, 4, 0, 3, 4, 9, 8, 7, 3, 13, 6],
>         [2, 8, 8, 10, 9, 2, 5, 4, 3, 0, 4, 6, 5, 4, 3, 9, 5],
>         [6, 13, 4, 14, 13, 7, 4, 8, 4, 4, 0, 10, 9, 8, 4, 13, 4],
>         [6, 7, 15, 6, 4, 9, 12, 5, 9, 6, 10, 0, 1, 3, 7, 3, 10],
>         [4, 5, 14, 7, 6, 7, 10, 4, 8, 5, 9, 1, 0, 2, 6, 4, 8],
>         [4, 8, 13, 9, 8, 7, 10, 3, 7, 4, 8, 3, 2, 0, 4, 5, 6],
>         [5, 12, 9, 14, 12, 6, 6, 7, 3, 3, 4, 7, 6, 4, 0, 9, 2],
>         [9, 10, 18, 6, 8, 12, 15, 8, 13, 9, 13, 3, 4, 5, 9, 0, 9],
>         [7, 14, 9, 16, 14, 8, 5, 10, 6, 5, 4, 10, 8, 6, 2, 9, 0],
>     ]
>     data['time_windows'] = [
>         (0, 5),  # depot
>         (7, 12),  # 1
>         (10, 15),  # 2
>         (16, 18),  # 3
>         (10, 13),  # 4
>         (0, 5),  # 5
>         (5, 10),  # 6
>         (0, 4),  # 7
>         (5, 10),  # 8
>         (0, 3),  # 9
>         (10, 16),  # 10
>         (10, 15),  # 11
>         (0, 5),  # 12
>         (5, 10),  # 13
>         (7, 8),  # 14
>         (10, 15),  # 15
>         (11, 15),  # 16
>     ]
>     data['num_vehicles'] = 6
>     data['depot'] = 0
>
>     # PICKUP AND DELIVERIES
>
>     #PICKUPS
>     data['demand'] = [0, 1, 1, -2, 4, 2, -4, 8, 8, 1, -2, 1, -2, 4, 4, -8, 8]
>
>     data['vehicle_capacities'] = [15, 15, 15, 15]
>
>
>     return data
>
>
> def print_solution(data, manager, routing, solution):
>     """Prints solution on console."""
>     print(f'Objective: {solution.ObjectiveValue()}')
>     time_dimension = routing.GetDimensionOrDie('Time')
>     total_time = 0
>     total_load = 0
>     for vehicle_id in range(data['num_vehicles']):
>         index = routing.Start(vehicle_id)
>         plan_output = 'Route for vehicle {}:\n'.format(vehicle_id)
>         route_load = 0
>         while not routing.IsEnd(index):
>             time_var = time_dimension.CumulVar(index)
>             node_index = manager.IndexToNode(index)
>             route_load += data['demand'][node_index]
>
>             plan_output += '{0} Time({1},{2}), Load {3} -> '.format( manager.IndexToNode(index), solution.Min(time_var), solution.Max(time_var), route_load)
>             index = solution.Value(routing.NextVar(index))
>         time_var = time_dimension.CumulVar(index)
>         plan_output += '{0} Time({1},{2}), Load {3}\n'.format(manager.IndexToNode(index),
>                                                     solution.Min(time_var),
>                                                     solution.Max(time_var), route_load)
>         plan_output += 'Time of the route: {}min\n'.format(
>             solution.Min(time_var))
>         print(plan_output)
>         total_time += solution.Min(time_var)
>     print('Total time of all routes: {}min'.format(total_time))
>
>
> def main():
>     """Solve the VRP with time windows."""
>     # Instantiate the data problem.
>     data = create_data_model()
>
>     # Create the routing index manager.
>     manager = pywrapcp.RoutingIndexManager(len(data['time_matrix']),
>                                            data['num_vehicles'], data['depot'])
>
>     # Create Routing Model.
>     routing = pywrapcp.RoutingModel(manager)
>
>     ##### ---- TIME ----
>     # Create and register a transit callback.
>     def time_callback(from_index, to_index):
>         """Returns the travel time between the two nodes."""
>         # Convert from routing variable Index to time matrix NodeIndex.
>         from_node = manager.IndexToNode(from_index)
>         to_node = manager.IndexToNode(to_index)
>         return data['time_matrix'][from_node][to_node]
>
>     transit_callback_index = routing.RegisterTransitCallback(time_callback)
>
>     # Define cost of each arc.
>     routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)
>
>
>
>
>
>     # Add Time Windows constraint.
>     time = 'Time'
>     routing.AddDimension(
>         transit_callback_index,
>         30,  # allow waiting time
>         30,  # maximum time per vehicle
>         False,  # Don't force start cumul to zero.
>         time)
>
>
>     time_dimension = routing.GetDimensionOrDie(time)
>     # Add time window constraints for each location except depot.
>     for location_idx, time_window in enumerate(data['time_windows']):
>         if location_idx == data['depot']:
>             continue
>         index = manager.NodeToIndex(location_idx)
>         time_dimension.CumulVar(index).SetRange(time_window[0], time_window[1])
>
>     # Add time window constraints for each vehicle start node.
>     depot_idx = data['depot']
>     for vehicle_id in range(data['num_vehicles']):
>         index = routing.Start(vehicle_id)
>         time_dimension.CumulVar(index).SetRange(
>             data['time_windows'][depot_idx][0],
>             data['time_windows'][depot_idx][1])
>
>     # Instantiate route start and end times to produce feasible times.
>     for i in range(data['num_vehicles']):
>         routing.AddVariableMinimizedByFinalizer(
>             time_dimension.CumulVar(routing.Start(i)))
>         routing.AddVariableMinimizedByFinalizer(
>             time_dimension.CumulVar(routing.End(i)))
>
>
>     ### PICKUPS AND
>     def demand_callback(from_index):
>         """Returns the demand of the node."""
>         # Convert from routing variable Index to demands NodeIndex.
>         from_node = manager.IndexToNode(from_index)
>         return data['demand'][from_node]
>
>
>     demand_callback_index = routing.RegisterUnaryTransitCallback(
>         demand_callback)
>
>
>     capacity= 'Capacity'
>     routing.AddDimension(
>         demand_callback_index,
>         0,  # null capacity slack
>         15,  # vehicle maximum capacities
>         False,  # start cumul to zero
>         capacity)
>
>
>     # Setting first solution heuristic.
>     search_parameters = pywrapcp.DefaultRoutingSearchParameters()
>     search_parameters.first_solution_strategy = (
>         routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC)
>
>     # Solve the problem.
>     solution = routing.SolveWithParameters(search_parameters)
>
>     # Print solution on console.
>     if solution:
>         print_solution(data, manager, routing, solution)
>
>
> if __name__ == '__main__':
>     main()
>
>
> Solution:
>
> Route for vehicle 0:
> 0 Time(0,0), Load 0 -> 0 Time(0,0), Load 0
> Time of the route: 0min
>
> Route for vehicle 1:
> 0 Time(0,0), Load 0 -> 0 Time(0,0), Load 0
> Time of the route: 0min
>
> Route for vehicle 2:
> 0 Time(0,0), Load 0 -> 9 Time(2,3), Load 1 -> 14 Time(7,8), Load 5 -> 16 Time(11,11), Load 13 -> 0 Time(18,18), Load 13
> Time of the route: 18min
>
> Route for vehicle 3:
> 0 Time(0,0), Load 0 -> 7 Time(2,4), Load 8 -> 1 Time(7,11), Load 9 -> 4 Time(10,13), Load 13 -> 3 Time(16,16), Load 11 -> 0 Time(24,24), Load 11
> Time of the route: 24min
>
> Route for vehicle 4:
> 0 Time(0,0), Load 0 -> 12 Time(4,4), Load -2 -> 13 Time(6,6), Load 2 -> 15 Time(11,11), Load -6 -> 11 Time(14,14), Load -5 -> 0 Time(20,20), Load -5
> Time of the route: 20min
>
> Route for vehicle 5:
> 0 Time(0,0), Load 0 -> 5 Time(3,3), Load 2 -> 8 Time(5,5), Load 10 -> 6 Time(7,7), Load 6 -> 2 Time(10,10), Load 7 -> 10 Time(14,14), Load 5 -> 0 Time(20,20), Load 5
> Time of the route: 20min
>
> Total time of all routes: 82min
>
>
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3074
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by SergiBago** [reply]
Hi James,

I already tried that option, but it didn't solve the problem.

Furthermore, I think I need the start cumul to be bigguer than 0, since the vehicles leave the depot (company headquarters) loaded with the employees, and start dropping them of. When the employers end the that work, they are picked up and moved to another location. At the end of the day, all employes must be picked up and leaved at de depot. 

All this conditions are okey with mi program, the only issue is the negative charge I mentioned above.

Thanks

**Comment by jmarca** [comment]
Then I suspect a mathematics error.

Make sure you set a hard condition on the starting load of each
vehicle to what you expect.  Then check your loads and unloads


On Thu, Jan 13, 2022 at 08:52:10AM -0800, Sergi Bagó Vila wrote:
> Hi James,
>
> I already tried that option, but it didn't solve the problem.
>
> Furthermore, I think I need the start cumul to be bigguer than 0, since the vehicles leave the depot (company headquarters) loaded with the employees, and start dropping them of. When the employers end the that work, they are picked up and moved to another location. At the end of the day, all employes must be picked up and leaved at de depot.
>
> All this conditions are okey with mi program, the only issue is the negative charge I mentioned above.
>
> Thanks
>
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3074#discussioncomment-1963922
> You are receiving this because you commented.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Comment by SergiBago** [comment]
I think that I need to add the condition to make sure that at every pickup/demand node, the load of the vehicle plus the demand of that node is >=0, but I tried several things and none of them worked. Could you please tell me how to do it? a piece of code would be a great help.

Thanks!

**Comment by jmarca** [comment]
On Thu, Jan 13, 2022 at 10:57:07AM -0800, Sergi Bagó Vila wrote:
> I think that I need to add the condition to make sure that at every pickup/demand node, the load of the vehicle plus the demand of that node is >=0, but I tried several things and none of them worked. Could you please tell me how to do it? a piece of code would be a great help.

No need, that should happen automatically.

If the vehicle has less than the required demand, unless there is slack
in the dimension, it cannot visit.

No, I don't have time to write code for you.

>
> Thanks!
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3074#discussioncomment-1964701
> You are receiving this because you commented.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Comment by Anshgala333** [comment]
i too have the same issue , where i try to deliver goods at a location 
i also added negative demands in the demand array , and the solutuion is not giving any solution

this is my demands array =  [0, 10, 30, -10, 30, 0] demands
this is my weight array = [0, 200, 300, -100, 600, 0] weight

i also added one extra coordinate for that specific location 

and also 
pickup_index = manager.NodeToIndex(2)   # delivery node index
    delivery_index = manager.NodeToIndex(3) # pickup node index

    routing.AddPickupAndDelivery(pickup_index, delivery_index)
    routing.solver().Add(
        routing.VehicleVar(pickup_index) == routing.VehicleVar(delivery_index)
    )
    routing.solver().Add(
        capacity_dimension.CumulVar(pickup_index) >= capacity_dimension.CumulVar(delivery_index)
    )


still doesnt work

**Reply by jmarca** [reply]
For this one, I think that you have to change the way you're approaching the problem.  A pickup and a delivery need to have paired demands as well.  In your case, I would have expected the positive 10 demand to be paired with delivery at the -10 demand location, but you say that you're instead linking nodes 2 and 3, which I believe have demands 30 and -10.  

You can also do what you're doing, and not pair up exact demands with exact supplies.  But, without sufficient slack value in the delivery  dimension, this is probably not going to work.  You did not post how you define your dimensions, so it is hard to guess what the problem is exactly. 

I always try to keep exact quantities paired up.  If there is a problem such as pouring coffee into three mugs, I would first model it as a pickup and delivery problem where there are three dummy nodes at the coffee pot location that supply one cup of coffee for each mug (or 100ml for one, 250ml for a second, and 350ml for a third, etc).  Once I had that working, I would that branch out to looser connections where the coffee maker has a large supply that can serve a variable number of mugs, etc etc.

Speaking of which, I'm  off to pickup another mug of coffee.

**Reply by Anshgala333** [reply]
yeah exactly , to give you a real scenario 
this is my demand array = [0 , 100 , 530 , 300 , -400 , 400]
this array means index 0 is depot or start 
index 1 deliver 100 boxes
index 2 deliver 530 box 
index 3 deliver 300 
this all are by default from depot

at index 4 pickup 400 box which will be delivered at index 5 
which i group in pickup and delivery -400 and +400 
index 4 and 5

and my max truck capacity is 630

**Reply by Anshgala333** [reply]
for your refernece this is the exact code 

from ortools.constraint_solver import routing_enums_pb2
from ortools.constraint_solver import pywrapcp
from utils.create_distance_matrix import create_distance_matrix
from config.logger import logger
from config.config import google_or_constant


async def create_data_model(all_data):
    data = {}
    coordinates = all_data["coordinates"]
    data["distance_matrix"] = await create_distance_matrix(coordinates)
    data["vehicle_capacities"] = all_data["truck_data"]["vehicle_capacities"]
    data["num_vehicles"] = all_data["truck_data"]["num_vehicles"]
    # data["demands"] = all_data["demands"]
    # data["weights"] = all_data["per_customer_total_weight"]
    data["demands"] = [0, 100, 530, 300, 400, -400, 0]
    data["weights"] = [0, 100, 530, 300, 400, -400, 0]
    data["vehicle_weights"] = all_data["vehicle_weight_capacity"]
    print(data["vehicle_capacities"])
    data["depot"] = 0
    return data




async def main(all_data):
    data = await create_data_model(all_data)
    logger.info("data model created for google or tools")
    end_index = len(all_data["coordinates"]) - 1

    manager = pywrapcp.RoutingIndexManager(
        len(data["distance_matrix"]),
        data["num_vehicles"],
        #  data["depot"]
        [data["depot"]] * data["num_vehicles"],
        [end_index] * data["num_vehicles"],
    )
    routing = pywrapcp.RoutingModel(manager)

    # -------------------------- distance constraint start ---------------------------

    # step 1:- creating a callback
    def distance_callback(from_index, to_index):
        from_node = manager.IndexToNode(from_index)
        to_node = manager.IndexToNode(to_index)
        return data["distance_matrix"][from_node][to_node]

    # step 2:- registering the callback
    transit_callback_index = routing.RegisterTransitCallback(distance_callback)

    # steo 3:- adding that dimension
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)
    routing.AddDimension(
        transit_callback_index,
        0,  # slack
        google_or_constant.MAX_DISTANCE_FOR_EACH_VEHICLE,  # max distance each vehicle can travel (example)
        True,  # start cumul to zero
        "Distance",
    )
    distance_dimension = routing.GetDimensionOrDie("Distance")
    # distance_dimension.SetGlobalSpanCostCoefficient(100)

    # -------------------------- distance constraint end -------------------------

    # ----------------- volume constraint start ----------------------------------------
    # step 1:- creating a demand callback
    def demand_callback(from_index):
        from_node = manager.IndexToNode(from_index)
        return data["demands"][from_node]

    # step 2:- registering that callback
    demand_callback_index = routing.RegisterUnaryTransitCallback(demand_callback)

    # step 3:- adding the volume dimension to the solver
    ### syntax :- (callback , )
    routing.AddDimensionWithVehicleCapacity(
        demand_callback_index, 0, data["vehicle_capacities"], True, "Capacity"
    )
    demand_dimension = routing.GetDimensionOrDie("Capacity")
    # demand_dimension.SetSpanCostCoefficientForAllVehicles(google_or_constant.PENALTY_FOR_VOLUME_UNDISTRIBUTION)
    # ----------------- volume constraint end ----------------------------------------

    # ----------------- weight constraint start ----------------------------------------
    def weight_callback(from_index):
        from_node = manager.IndexToNode(from_index)
        return data["weights"][from_node]

    weight_callback_index = routing.RegisterUnaryTransitCallback(weight_callback)
    routing.AddDimensionWithVehicleCapacity(
        weight_callback_index, 0, data["vehicle_weights"], True, "Weight"  # no slack
    )

    weight_dimension = routing.GetDimensionOrDie("Weight")
    # weight_dimension.SetSpanCostCoefficientForAllVehicles(google_or_constant.PENALTY_FOR_WEIGHT_UNDISTRIBUTION)

    for vehicle_id in range(data["num_vehicles"]):
        end_index = routing.End(vehicle_id)
        weight_dimension.CumulVar(end_index).SetMax(data["vehicle_weights"][vehicle_id])

    # ----------------- weight constraint end ----------------------------------------

    data["pickups_deliveries"] = [
        [4, 5],
    ]

    for request in data["pickups_deliveries"]:
        pickup_index = manager.NodeToIndex(request[0])
        delivery_index = manager.NodeToIndex(request[1])
        routing.AddPickupAndDelivery(pickup_index, delivery_index)
        routing.solver().Add(
            routing.VehicleVar(pickup_index) == routing.VehicleVar(delivery_index)
        )
        routing.solver().Add(
            distance_dimension.CumulVar(pickup_index)
            < distance_dimension.CumulVar(delivery_index)
        )

    # Force same vehicle for pickup and delivery, and delivery before pickup
    # pickup_index = manager.NodeToIndex(2)   # delivery node index
    # delivery_index = manager.NodeToIndex(3) # pickup node index

    # routing.AddPickupAndDelivery(pickup_index, delivery_index)
    # routing.solver().Add(
    #     routing.VehicleVar(pickup_index) == routing.VehicleVar(delivery_index)
    # )
    # routing.solver().Add(
    #         distance_dimension.CumulVar(pickup_index)
    #         <= distance_dimension.CumulVar(delivery_index)
    #     )

    search_parameters = pywrapcp.DefaultRoutingSearchParameters()
    search_parameters.first_solution_strategy = (
        routing_enums_pb2.FirstSolutionStrategy.PATH_CHEAPEST_ARC
    )
    search_parameters.local_search_metaheuristic = (
        routing_enums_pb2.LocalSearchMetaheuristic.GUIDED_LOCAL_SEARCH
    )
    search_parameters.time_limit.FromSeconds(5)

    solution = routing.SolveWithParameters(search_parameters)

    if solution:
        logger.info("google or tools gave a solution")
        structured_data = generate_output(
            data, manager, routing, solution, all_data["customer_names"], all_data
        )
        return structured_data
    else:
        print("no solution")
        logger.error("No solution were found from google or")


if __name__ == "__main__":
    main()

**Reply by jmarca** [reply]
Okay, that looks impossible to me.

If the truck is loaded with 630 at the depot, it can deliver to node 1, then node 2, refresh another 400 at node 4, deliver 400 at 5, but then it is going to drop node 3.

Unless you have a disjunction allowing the solver to drop nodes, this will fail.

**Reply by Anshgala333** [reply]
this is the output 

🚛 Truck 1 → Capacity: 9000 kg
📦 Total Boxes: 530 | 🏋‍♂ Total Weight: 500 kg
📍 Route (Customer Indexes): [0, 2]
🧍 Customers Served: ['Raj']
───────

🚛 Truck 2 → Capacity: 9000 kg
📦 Total Boxes: 800 | 🏋‍♂ Total Weight: 400 kg
📍 Route (Customer Indexes): [0, 1, 4, 5, 3]
🧍 Customers Served: ['Ansh', 'Raj', 'Harsh', 'dummy']
───────

but somehow exceed the truck capcity

**Reply by Anshgala333** [reply]
okay forgot to added some location are added twice like x box need to delivered and y box need to be picked which will be delievred at z location 
and y and z are grouped

**Reply by jmarca** [reply]
Allow disjunctions at nodes, with a high cost so that the solver does not want to drop the nodes.

**Reply by Anshgala333** [reply]
but as of now it is not dropping any nodes , instead taking all

**Reply by jmarca** [reply]
And if you have multiple visits per location, you need to generate multiple nodes for those locations, because each node in the model can only be visited once.

**Reply by Anshgala333** [reply]
my only problem is a valid solution wrt distance , such that all orders are delivered which start from depot and also those whose pickup and drop are decided , and also not exceed truck capacity

**Reply by Anshgala333** [reply]
yes i do "And if you have multiple visits per location, you need to generate multiple nodes for those locations, because each node in the model can only be visited once."

**Reply by Anshgala333** [reply]
what do you think what approach would work here 
in official google or tools documentation for pickup and drop , they also only provide routes , not in constraints of demand and truck capacity
or thats just not possible ..?
